### PR TITLE
New configurable/overridable kernel ZMQ+Websocket connection API

### PR DESCRIPTION
**Pull request by Zsailer**
## Summary

Decouples the ZMQ socket handling logic from the websocket handler definition to enable server deployers to override this logic without forking the websocket handler.

## The Problem

The logic for connecting a websocket to a kernel's set of ZMQ sockets is defined under methods in kernel's websocket handler. 

This presents the following problem:

If you need to change, override, and remove logic underneath the kernel's `/api/kernels/<kernel-id>/channels` handler, you are forced to 1) fork Jupyter Server's `ZMQChannelshandler` and 2) insert the forked version earlier in the list of Jupyter Server's handlers to ensure it intercepts all kernel channels requests.

(This PR was inspired by some work I have been doing, experimenting with adding _events_ and additional logging to the kernel's websocket API and found it impossible to do this without forking.)

## Proposed solution

Separate the "business" logic into an overrideable, configurable class.

In other services, Jupyter Server uses a different pattern than the current websocket handler; it decouples "business" logic from the REST API handlers/definitions by defining a configurable, overrideable "manager" that follows a defined interface and handles the "business" logic underneath the REST API handlers. You can completely override the logic in the managers, as long as they follow the defined manager interface (usually an ABC).

In this PR, I've defined a new abstraction/interface called a `KernelWebsocketConnectionABC`. The interface is minimal and clearly maps to the methods in Tornado's default `WebsocketHandler`. 

I've also added a base class that handles the instantiation of Jupyter Client Session objects.

Then, the default implementation, called `ZMQChannelsWebsocketConnection` includes all the methods previously found in the websocket handler. 

## A Note on Backwards Compatibility.

Most user/deployers will not be impacted by this change.

This impacts anyone who has forked the `ZMQChannelsHandler`, but it doesn't _break_ their fork. Presumably, anyone who has forked this handler will still inject their handler _before_ this new API is ever called. As a result, this change would leave them unaffected, but we should recommend that they update their code to use this new API.

## Todo

- [x] Make `prepare` optional
- [x] Make `open` async
- [x] Make `parent` of the kernel websocket connection object the `kernel_manager` and `parent.parent` the `multi_kernel_manager`.
- [x] deprecate traits on `ServerApp` and move to `ZMQChannelsWebsocketConnection`
- [x] Get tests passing

### CODE REVIEW

**Comment by blink1073** [comment]
Notes from today's discussion

- `prepare` could be made optional
- `open` can be async
- `parent` could be the KernelManager

**Comment by blink1073** [comment]
The macos failures are unrelated, I'll make them more robust in #1069.  There are a number of typing failures though.

**Comment by Zsailer** [comment]
Bringing this out of draft state. I think it is ready for final review. 

@blink1073  do you mind taking a look over?

File: tests/services/sessions/test_api.py [hunk 1]
```diff
@@ -556,9 +556,13 @@ async def test_restart_kernel(session_client, jp_base_url, jp_fetch, jp_ws_fetch
     model = json.loads(r.body.decode())
     assert model["connections"] == 0
 
-    # Open a websocket connection.
-    await jp_ws_fetch("api", "kernels", kid, "channels")
+    # Add a delay to give the kernel enough time to restart.
+    #    time.sleep(2)
```
Comment (Zsailer) [inline tests/services/sessions/test_api.py:560]:
```suggestion
```

**Comment by blink1073** [comment]
Yep, I'll take a look, but probably not today.

File: jupyter_server/base/zmqhandlers.py [hunk 2]
```diff
@@ -1,350 +1,9 @@
...
-        res = super().get(*args, **kwargs)
-        await res
-
-    def initialize(self):
-        self.log.debug("Initializing websocket connection %s", self.request.path)
-        self.session = Session(config=self.config)
-
-    def get_compression_options(self):
-        return self.settings.get("websocket_compression_options", None)
+"""Add deprecation warning here.
```
Comment (blink1073) [inline jupyter_server/base/zmqhandlers.py:1]: Still TODO?

File: jupyter_server/serverapp.py [hunk 3]
```diff
@@ -141,6 +120,51 @@
...
+
+try:
+    import tornado
+
+    assert tornado.version_info >= MIN_TORNADO
+except (ImportError, AttributeError, AssertionError) as e:  # pragma: no cover
+    raise ImportError(_i18n("The Jupyter Server requires tornado >=%s.%s.%s") % MIN_TORNADO) from e
+
+
+try:
```
Comment (blink1073) [inline jupyter_server/serverapp.py:146]: This block is repeated from above

File: jupyter_server/services/kernels/connection/channels.py [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+                await ready
+            except Exception as e:
+                self.kernel_manager.execution_state = "dead"
+                self.kernel_manager.reason = str(e)
+                raise web.HTTPError(500, str(e)) from e
+
+        t0 = time.time()
+        while not await ensure_async(self.kernel_manager.is_alive()):
+            await asyncio.sleep(0.1)
+            if time.time() - t0 > self.multi_kernel_manager.kernel_info_timeout:
```
Comment (blink1073) [inline jupyter_server/services/kernels/connection/channels.py:311]: I think this needs parens around the left hand side of the comparison operator

**Review by blink1073 — CHANGES_REQUESTED** [review]

File: jupyter_server/serverapp.py (hunk 3 above)
Comment (Zsailer) [reply jupyter_server/serverapp.py:149]: I have no idea how that happened... Fixed 😆

File: jupyter_server/base/zmqhandlers.py (hunk 2 above)
Comment (Zsailer) [reply jupyter_server/base/zmqhandlers.py]: Ah, yes. I've now added a deprecation warning for this module.

File: jupyter_server/services/kernels/connection/channels.py (hunk 4 above)
Comment (Zsailer) [reply jupyter_server/services/kernels/connection/channels.py]: Ah right :+1: Done.

**Comment by Zsailer** [comment]
The Jupyter Server Terminals tests are failing because of the deprecation warning on the zmqhandlers module. I can remove this warning if preferred, or update jupyter_server_terminals to import from the new location, `jupyter_server.services.kernels.websocket`. What do you think?

**Comment by blink1073** [comment]
Since the terminals plugin depends on 2.0, I'd say update there

**Comment by blink1073** [comment]
But not until after the rc is released

**Comment by blink1073** [comment]
Actually, it seems like we need to ignore the warning here, release the RC, release a fix in terminals, and then remove the warning ignore here.

**Comment by Zsailer** [comment]
Okay—I've ignored the warnings here. The downstream tests will still fail because we need to update jupyter_server terminals. I've opened this [PR in jupyter_server_terminals](https://github.com/jupyter-server/jupyter_server_terminals/pull/62) to handle the new module.

If I understand correctly, we'll 
1. merge this PR even though the downstream tests are failing
2. cut an RC with these changes
3. kick CI in my PR in jupyter_server_terminals to pick up the new Jupyter Server rc.
4. when it passes, merge that PR and cut a release of jupyter_server_terminals.

**Comment by blink1073** [comment]
Yep, sounds good!

**Comment by blink1073** [comment]
Okay, I'll cut an RC tomorrow with this.  We can still target 5 Dec for the final since a lot of folks will be out of office this week.

**Review by blink1073 — APPROVED** [review]
Thanks!

**Comment by bloomsa** [comment]
I was _just_ looking at overriding the /<kernel-id>/channels endpoint to do a check that the user requesting the WS was the user that owned the kernel. great timing on this abstraction piece! thank you

**Comment by bloomsa** [comment]
> I was _just_ looking at overriding the /kernel-id/channels endpoint to do a check that the user requesting the WS was the user that owned the kernel. great timing on this abstraction piece! thank you

shoot, I was hoping to use this with Enterprise-gateway but conflicting deps. force me to use a 1.x version of jupyter-server. Is a backport of this to 1.x possible, or does it depend on other 2.0 functionality? @Zsailer (not asking you to do the work, I can try to implement it, just wondering about the feasibility)

**Comment by kevin-bates** [comment]
Hi @bloomsa.  (I am "out of the office" this week, as, I believe, is @Zsailer - just fyi.)

My understanding is that this feature is intended for 2.0+.  Should a backport to 1.x be considered, it **must not** introduce a dependency on jupyter_client 7+ (which is currently the reason EG needs to cap Server < 2)  since current EG releases (<= 3.0) require jupyter_client < 7.  That is, the _dependency ice_ on which EG rests is extremely thin! :smile:

I think it may make more sense to use this feature (which I agree is great!) as another impetus to getting EG 4.0 going where we plan to derive EG directly from `ServerApp` along with [_Remote Provisioners_](https://github.com/gateway-experiments/remote_provisioners) to continue supporting resource-managed kernels.  Derivation from `ServerApp` would then enable EG to support things like server extensions etc.  We could then _repurpose_ EG functionality like [_Kernelspec Caching_](https://github.com/jupyter-server/enterprise_gateway/pull/908), [_Dynamic Configurables_](https://github.com/jupyter-server/enterprise_gateway/pull/748), and applicable portions of the [_High Availability_](https://github.com/jupyter-server/enterprise_gateway/pull/1095) functionality as server extensions while moving us closer to [_service composability_](https://github.com/jupyter-server/jupyter_server/issues/178).

**Comment by bloomsa** [comment]
@kevin-bates I appreciate the reply, especially while OOO! My use case for this is to ensure that the person trying to connect to a kernel with the `api/kernels/kernel-id/channels` is the same as the user that created the kernel in the first place. Preventing a situation where someone can access a kernel and its state when they aren't supposed to. I _think_ I am targeting the right spot, being the ZMQChannelsHandler, with overriding its `get` mapping method to validate a user. LMK if this is something the Jupyter team has already considered or if another approach would make more sense

**Comment by kevin-bates** [comment]
I'm not sure how much something like this will interfere with RTC (real-time collaboration) aside from my concern about EG's precarious dependency position - so I'm more concerned about _when and where_ than how.  I think it's great that you're looking into this, but we should also keep in mind the other areas of recent activity like RTC and authorization.  Unfortunately, I'm not intimately familiar with RTC or the authorization work (I apologize).

I'm assuming your changes would be directed at the EG server itself, not the EG gateway client package built into ServerApp - correct?

Once EG 4.0 is in place (deriving from or at least closer to `ServerApp` and no ceiling on jupyter_client) we would be in a better position to leverage these kinds of things.

**Comment by bloomsa** [comment]
@kevin-bates I hadn't considered the interference with RTC, good point. In the end, I was able to override the ZMQChannelsHandler method simply by making a child class off of ZMQChannelsHandler and overriding the `get` method in my Enterprise Gateway implementation. It worked similar to how the kernel handlers are overridden with Mixins right now. So there was no need for this to be backported after all 👍

**Comment by kevin-bates** [comment]
Great -thanks for getting back to us @bloomsa.

**Comment by echarles** [comment]
> I think it may make more sense to use this feature (which I agree is great!) as another impetus to getting EG 4.0 going where we plan to derive EG directly from ServerApp along with [Remote Provisioners](https://github.com/gateway-experiments/remote_provisioners)

Is there an issue that tracks EG on ServerApp + RemoteProvisionner?

**Comment by kevin-bates** [comment]
@echarles - It's _vaguely_ on the EG [project roadmap](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/contributors/roadmap.html#planned-for-4-0) but it's probably time we can open an issue or two (in EG).

**Comment by echarles** [comment]
@kevin-bates https://github.com/jupyter-server/enterprise_gateway/issues/1208

**Pull request merged**